File: mill-example/oystercard/src/oystercard/scala/TubeFare.scala
package oystercard.scala

/** Represents a tube fare rule
  *
  * @param name the rule name (e.g. 'Anywhere in Zone 1')
  * @param price the fare cost
  * @param applies a predicate which checks if the fare is applicable
  */
final case class TubeFare private(name: String, price: Amount, applies: TubeFare.Applies)

object TubeFare {
  type FromZones = Set[Zone]
  type ToZones = Set[Zone]
  type Applies = (FromZones, ToZones) => Boolean
}

File: mill-example/oystercard/src/oystercard/java/TubeFare.java
package oystercard.java;

import java.util.function.BiFunction;
import java.util.Set;

public final class TubeFare {
    public final String ruleName;
    public final Amount price;

    public int getPriceInPence() {
        return price.valueInPence;
    }
    public final BiFunction<Set<Zone>, Set<Zone>, Boolean> ruleApplies;
    public TubeFare(String  name, Amount prc, BiFunction<Set<Zone>, Set<Zone>, Boolean> applies) {
        ruleName = name;
        price = prc;
        ruleApplies = applies;
    }
    public String toString() {
        return ruleName + " is " + price;
    }
}
